### PR TITLE
Fix optional trait parsing

### PR DESCRIPTION
**Pull request by dme2**
Addresses #2725

### CODE REVIEW

**Comment by powerboat9** [comment]
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0363d352bef03bd91009d58c499ebf95

File: gcc/rust/parse/rust-parse-impl.h [hunk 1]
```diff
@@ -4939,6 +4939,13 @@ Parser<ManagedTokenSource>::parse_trait (AST::Visibility vis,
     {
       lexer.skip_token ();
 
+      if (lexer.peek_token()->get_id() == QUESTION_MARK) {
+            lexer.skip_token();
+            rust_error_at(lexer.peek_token()->get_locus(),
+                          "%<?Trait%> is not permitted in supertraits");
+            return nullptr;
+      }
```
Comment (CohenArthur) [inline gcc/rust/parse/rust-parse-impl.h:4947]: yes, as @powerboat9 showed we should not be emitting this error at the parsing stage, but probably in the typechecker. you can see that it parses properly in the playground link, or with a nightly `rustc` with the `-Z parse-only` CLI option

Comment (CohenArthur) [reply gcc/rust/parse/rust-parse-impl.h:4947]: if you'd like any guidance for adding that check to the type checker please do say so :)

Comment (dme2) [reply gcc/rust/parse/rust-parse-impl.h:4947]: Thanks for the feedback. I'll go ahead and try to add the check to the typechecker.

**Comment by dme2** [comment]
Hi @CohenArthur, I know the formatting isn't passing yet, but does that latest commit look ok?

File: gcc/rust/typecheck/rust-hir-type-check-item.cc [hunk 2]
```diff
@@ -518,6 +518,20 @@ TypeCheckItem::visit (HIR::Module &module)
 void
 TypeCheckItem::visit (HIR::Trait &trait)
 {
+  if (trait.has_type_param_bounds ())
+    {
+      for (auto &tp_bound : trait.get_type_param_bounds ())
+        {
+          HIR::TraitBound &tb
+              = static_cast<HIR::TraitBound &> (*tp_bound.get ());
+          if (tb.get_polarity () == BoundPolarity::AntiBound)
```
Comment (CohenArthur) [inline gcc/rust/typecheck/rust-hir-type-check-item.cc:527]: this is good, but the static cast might fail if the trait bound is a lifetime.

you can switch on `tp_bound.get_bound_type()` and check if it is a lifetime or a trait, and proceed to cast safely then. so something like:
```suggestion
          if (tp_bound.get_bound_type() == LIFETIME) {
          HIR::TraitBound &tb
              = static_cast<HIR::TraitBound &> (*tp_bound.get ());
          if (tb.get_polarity () == BoundPolarity::AntiBound)
          /* snip */
          }
```

etc. does that make sense?

Comment (dme2) [reply gcc/rust/typecheck/rust-hir-type-check-item.cc:527]: @CohenArthur, thank you for the review. Do  you have an example of the when the static cast might fail?

This program seems to compile fine with the code as is:

```rust
#[lang = "sized"]
pub trait Sized {}
trait Trait<'a>: 'a {}
```

Comment (CohenArthur) [reply gcc/rust/typecheck/rust-hir-type-check-item.cc:527]: ah, my bad - the static cast won't fail, but the value will be undefined. so we still need to do the check.

**Comment by P-E-P** [comment]
@dme2 What is the current state of this PR ? From what I've read it seems like there is only a minor format issue ?

**Comment by P-E-P** [comment]
Closed in favor of #3055 (This is exactly the same PR with proper code format). Thank you for the initial work!

**Pull request closed without merging**